### PR TITLE
fix(linux): Don't start ibus-daemon multiple times 🚌

### LINKED ISSUE #6237: bug(linux): keyboard installation starts additional `ibus-daemon`
During acceptance tests ([SUITE_INSTALLATION](https://github.com/keymanapp/keyman/issues/6200#issuecomment-1036315429)) the keyboards stopped working at one point. In some cases this was caused by a crashing ibus, in other cases by having two `ibus-daemon` processes.

### PR DESCRIPTION
**Pull request by ermshiperete**
This change adds a 1s sleep after restarting ibus before checking that it actually runs. Hopefully this will solve the problem. Since
this bug is hard to reproduce, we also log a critical error for the other code path this could happen.

Fixes #6237.

# User Testing

## Preparations

- Since this bug is hard to reliably reproduce, the tests should be run on all Linux platforms:
  - GROUP_BIONIC: Ubuntu 18.04 Bionic with Gnome Shell and X11
  - GROUP_FOCAL: Ubuntu 20.04 Focal with Gnome Shell and X11
  - GROUP_IMPISH_X11: Ubuntu 21.10 Impish with Gnome Shell and X11
  - GROUP_IMPISH_WAYLAND: Ubuntu 21.10 Impish with Gnome Shell and Wayland
  - GROUP_WASTA: Wasta 20.04 with Cinnamon

- [Install build artifacts of this PR](https://github.com/keymanapp/keyman/wiki/How-to-test-artifacts-from-pull-requests-for-Keyman-for-Linux)
- Reboot

## Tests

- **TEST_KB_INSTALL1**: Download and install a keyboard
  - On the command line, run:
    ```bash
    km-config -v 2>&1 | tee /tmp/km-config.log
    ```
  - click "Download", enter search for "Khmer", and install the "Khmer Angkor" keyboard
  - On the command line, run:
    ```bash
    ps --user $USER -o cmd -o pid= -o s= | grep -e "^ibus-daemon"
    ```
  - Verify that this outputs exactly one line starting with `ibus-daemon`
  - If the test fails, attach the file `/tmp/km-config.log` or paste the content of `/tmp/km-config.log`

- **TEST_KB_INSTALL2**: Download and install a second keyboard
  - repeat the previous test with a different keyboard:
  - On the command line, run:
    ```bash
    km-config -v 2>&1 | tee /tmp/km-config2.log
    ```
  - click "Download", enter search for "Icelandic", and install the "EuroLatin (SIL)" keyboard
  - On the command line, run:
    ```bash
    ps --user $USER -o cmd -o pid= -o s= | grep -e "^ibus-daemon"
    ```
  - Verify that this outputs exactly one line starting with `ibus-daemon`
  - If the test fails, attach the file `/tmp/km-config2.log` or paste the content of `/tmp/km-config2.log`

### CODE REVIEW

**Review by darcywong00 — APPROVED** [review]
lgtm

**Comment by bharanidharanj** [comment]
* **GROUP_FOCAL**: **Ubuntu 20.04 Focal with Gnome Shell and X11**

   * **TEST_KB_INSTALL1 (PASSED):** The output line starting with ibus-daemon.

     

   * **TEST_KB_INSTALL2 (PASSED):** The output line starting with ibus-daemon.

**Comment by bharanidharanj** [comment]
* **GROUP_BIONIC**: **Ubuntu 18.04 Bionic with Gnome Shell and X11**

   * **TEST_KB_INSTALL1 (PASSED):** The output line starting with ibus-daemon.

       

   * **TEST_KB_INSTALL2 (FAILED):** The output message starting with ibus-daemon with double lines.

**Comment by bharanidharanj** [comment]
* **GROUP_IMPISH_X11**: **Ubuntu 21.10 Impish with Gnome Shell and X11**

   * **TEST_KB_INSTALL1 (FAILED):** The output message starting with ibus-daemon with double lines.

        

   * **TEST_KB_INSTALL2 (FAILED):** The output message starting with ibus-daemon with double lines.

**Comment by bharanidharanj** [comment]
* **GROUP_IMPISH_WAYLAND**: **Ubuntu 21.10 Impish with Gnome Shell and Wayland**

   * **TEST_KB_INSTALL1 (PASSED):** Tested this on Ubuntu on Xorg and the output line starting with ibus-daemon.

            


   * **TEST_KB_INSTALL2 (PASSED):** Tested this on Ubuntu on Xorg and the output line starting with ibus-daemon.

File: linux/keyman-config/keyman_config/ibus_util.py [hunk 1]
```diff
@@ -130,6 +132,7 @@ def restart_ibus(bus=None):
         except Exception as e:
             logging.warning("Failed to restart IBus")
             logging.warning(e)
+    time.sleep(1)  # 1s
```
Comment (mcdurdin) [inline linux/keyman-config/keyman_config/ibus_util.py:135]: Can you add a comment explaining this delay?

**Comment by ermshiperete** [comment]
@bharanidharanj Could I ask you to re-run the tests with the new instructions. Thanks! I added some debug output, so hopefully this will allow me to figure out what's going wrong. In my local testing I'm not able to reproduce the problem after having the additional debug statements in place.

@keymanapp-test-bot retest all

File: linux/keyman-config/keyman_config/ibus_util.py (hunk 1 above)
Comment (ermshiperete) [reply linux/keyman-config/keyman_config/ibus_util.py:141]: done

**Review by mcdurdin — APPROVED** [review]

**Comment by bharanidharanj** [comment]
### GROUP_BIONIC: Ubuntu 18.04 Bionic with Gnome Shell and X11

* **TEST_KB_INSTALL1 (FAILED):** Tested this in Bionic 18.04 Linux OS and this test is failed. Attached km-config.log file for reference. 


 
[km-config.log](https://github.com/keymanapp/keyman/files/8212244/km-config.log)

**Comment by bharanidharanj** [comment]
### GROUP_BIONIC: Ubuntu 18.04 Bionic with Gnome Shell and X11

* **TEST_KB_INSTALL2 (FAILED):** This test is failed in Bionic OS. Attached the log file for reference. 

 



[km-config2.log](https://github.com/keymanapp/keyman/files/8212839/km-config2.log)

**Comment by bharanidharanj** [comment]
> 



> @bharanidharanj Could I ask you to re-run the tests with the new instructions. Thanks!
> 
> I added some debug output, so hopefully this will allow me to figure out what's going wrong. In my local testing I'm not able to reproduce the problem after having the additional debug statements in place.
> 
> @keymanapp-test-bot retest all

@ermshiperete  Okay, I will re-run the tests with the new instructions, you have mentioned in this PR.

**Comment by bharanidharanj** [comment]
### GROUP_FOCAL: Ubuntu 20.04 Focal with Gnome Shell and X11

* **TEST_KB_INSTALL1 (PASSED):** This Test is Passed with New Instructions.
* **TEST_KB_INSTALL2 (PASSED):** This Test is Passed with New Instructions. However, I got crash after installing the Keyman Keyboard. Attached it for your reference.

**Comment by bharanidharanj** [comment]
### GROUP_IMPISH_X11: Ubuntu 21.10 Impish with Gnome Shell and X11

* **TEST_KB_INSTALL1 (PASSED):** Tested this in Ubuntu 21.10 and it is working as expected.
* **TEST_KB_INSTALL2 (PASSED):** Tested this in Ubuntu 21.10 and it is working as expected.

### GROUP_IMPISH_WAYLAND: Ubuntu 21.10 Impish with Gnome Shell and Wayland

* **TEST_KB_INSTALL1 (PASSED):** Tested this in Ubuntu 21.10 and it is working as expected.
* **TEST_KB_INSTALL2 (PASSED):** Tested this in Ubuntu 21.10 and it is working as expected.

**Comment by bharanidharanj** [comment]
### GROUP_WASTA: Wasta 20.04 with Cinnamon

* **TEST_KB_INSTALL1 (PASSED):** Tested this in Wasta 20.04 Linux OS and it is working as expected.
* **TEST_KB_INSTALL2 (PASSED):** Tested this in Wasta 20.04 Linux OS and it is working as expected.

**Comment by bharanidharanj** [comment]
### GROUP_BIONIC: Ubuntu 18.04 Bionic with Gnome Shell and X11

* **TEST_KB_INSTALL1 (FAILED):** Tested this in Bionic Linux and the output shows more than one line. 

[km-config.log](https://github.com/keymanapp/keyman/files/8214333/km-config.log)


* **TEST_KB_INSTALL2 (FAILED):** Tested this in Bionic Linux and the output shows more than one line. 

[km-config2.log](https://github.com/keymanapp/keyman/files/8214334/km-config2.log)

**Comment by ermshiperete** [comment]
@bharanidharanj Can you please test again? Don't forget to install the new build (linked in the ["User Test Results" comment](https://github.com/keymanapp/keyman/pull/6283#issuecomment-1049742056)). Please test on all platforms, starting with Bionic, until you get a failure. Thanks!

@keymanapp-test-bot retest GROUP_BIONIC all

**Comment by bharanidharanj** [comment]
> @bharanidharanj Can you please test again? Don't forget to install the new build (linked in the ["User Test Results" comment](https://github.com/keymanapp/keyman/pull/6283#issuecomment-1049742056)). Please test on all platforms, starting with Bionic, until you get a failure. Thanks!
> 
> @keymanapp-test-bot retest GROUP_BIONIC all

@ermshiperete Sure. I will test it and will post my comment.

**Comment by bharanidharanj** [comment]
### GROUP_BIONIC: Ubuntu 18.04 Bionic with Gnome Shell and X11

* **TEST_KB_INSTALL1 (FAILED):** Re-tested this in Bionic and I getting more than 1 outputs in the Terminal. Attached the log file for reference. 

[km-config.log](https://github.com/keymanapp/keyman/files/8222045/km-config.log)

* **TEST_KB_INSTALL2 (FAILED):** Re-tested this in Bionic and I am getting more than 1 outputs in the Terminal. Attached the log file for reference. 

[km-config2.log](https://github.com/keymanapp/keyman/files/8222046/km-config2.log)

**Comment by ermshiperete** [comment]
@bharanidharanj That doesn't look like it's running the version from this PR. Can you show me the output of `km-config --version`?

**Comment by bharanidharanj** [comment]
> 

@ermshiperete km-config version 16.0.2-alpha

**Comment by ermshiperete** [comment]
> @ermshiperete km-config version 16.0.2-alpha

@bharanidharanj That should be 15.0.210-beta. Please install the package built for this PR by following the steps in the [wiki](https://github.com/keymanapp/keyman/wiki/How-to-test-artifacts-from-pull-requests-for-Keyman-for-Linux). Please let me know if the instructions in the wiki are not clear enough. Thanks!

**Comment by bharanidharanj** [comment]
> 

@ermshiperete As per our discussion, I have attached the Screenshot as well as the log file for your reference. 

 

[km-config.log](https://github.com/keymanapp/keyman/files/8231359/km-config.log)

**Comment by ermshiperete** [comment]
@bharanidharanj Please try again. I made some preliminary changes - so you can stop testing after the first failure.

@keymanapp-test-bot retest all

**Comment by MakaraSok** [comment]
# Test Results

### GROUP_BIONIC: Ubuntu 18.04 Bionic with Gnome Shell and X11

* **TEST_KB_INSTALL1 (FAILED):** 

I've uninstalled and removed everything related to Keyman using 
```
sudo apt remove keyman ibus-keyman python3-keyman-config libkmnkbp0-0
sudo apt autoremove
```
and 
```
rm -rf ~/.local/share/keyman/
sudo rm -rf /usr/local/share/keyman/
```

then after restarting the machine, the instructions at https://github.com/keymanapp/keyman/wiki/How-to-test-artifacts-from-pull-requests-for-Keyman-for-Linux were followed, but the second line of commands show some suspicious errors, see below: 

```
keyman@keyman-VirtualBox:~/Desktop/archive-6/results$ sudo apt update
[sudo] password for keyman: 
Hit:1 http://kh.archive.ubuntu.com/ubuntu bionic InRelease
Hit:2 http://kh.archive.ubuntu.com/ubuntu bionic-updates InRelease             
Hit:3 http://kh.archive.ubuntu.com/ubuntu bionic-backports InRelease           
Hit:4 http://security.ubuntu.com/ubuntu bionic-security InRelease              
Reading package lists... Done
Building dependency tree       
Reading state information... Done
1 package can be upgraded. Run 'apt list --upgradable' to see it.
keyman@keyman-VirtualBox:~/Desktop/archive-6/results$ sudo dpkg -i \
> {ibus-keyman,libkmnkbp0}*$(lsb_release -s -c)*$(dpkg --print-architecture).deb \
> {keyman,python3-keyman-config}*$(lsb_release -s -c)*all.deb
Selecting previously unselected package ibus-keyman.
(Reading database ... 168797 files and directories currently installed.)
Preparing to unpack ibus-keyman_15.0.212-1~PR-6283+bionic21_amd64.deb ...
Unpacking ibus-keyman (15.0.212-1~PR-6283+bionic21) ...
Selecting previously unselected package libkmnkbp0-0:amd64.
Preparing to unpack libkmnkbp0-0_15.0.212-1~PR-6283+bionic21_amd64.deb ...
Unpacking libkmnkbp0-0:amd64 (15.0.212-1~PR-6283+bionic21) ...
Selecting previously unselected package keyman.
Preparing to unpack keyman_15.0.212-1~PR-6283+bionic21_all.deb ...
Unpacking keyman (15.0.212-1~PR-6283+bionic21) ...
Selecting previously unselected package python3-keyman-config.
Preparing to unpack python3-keyman-config_15.0.212-1~PR-6283+bionic21_all.deb ...
Unpacking python3-keyman-config (15.0.212-1~PR-6283+bionic21) ...
Setting up libkmnkbp0-0:amd64 (15.0.212-1~PR-6283+bionic21) ...
dpkg: dependency problems prevent configuration of python3-keyman-config:
 python3-keyman-config depends on python3-bs4; however:
  Package python3-bs4 is not installed.
 python3-keyman-config depends on python3-sentry-sdk (>= 1.1) | python3-raven; however:
  Package python3-sentry-sdk is not installed.
  Package python3-raven is not installed.
 python3-keyman-config depends on python3-lxml; however:
  Package python3-lxml is not installed.
 python3-keyman-config depends on python3-magic; however:
  Package python3-magic is not installed.
 python3-keyman-config depends on python3-numpy; however:
  Package python3-numpy is not installed.
 python3-keyman-config depends on python3-qrcode; however:
  Package python3-qrcode is not installed.
 python3-keyman-config depends on python3-requests-cache; however:
  Package python3-requests-cache is not installed.

dpkg: error processing package python3-keyman-config (--install):
 dependency problems - leaving unconfigured
Setting up ibus-keyman (15.0.212-1~PR-6283+bionic21) ...
dpkg: dependency problems prevent configuration of keyman:
 keyman depends on python3-keyman-config (= 15.0.212-1~PR-6283+bionic21); however:
  Package python3-keyman-config is not configured yet.

dpkg: error processing package keyman (--install):
 dependency problems - leaving unconfigured
Processing triggers for man-db (2.8.3-2ubuntu0.1) ...
Processing triggers for libc-bin (2.27-3ubuntu1.5) ...
Processing triggers for gnome-menus (3.13.3-11ubuntu1.1) ...
Processing triggers for desktop-file-utils (0.23-1ubuntu3.18.04.2) ...
Processing triggers for mime-support (3.60ubuntu1) ...
Processing triggers for libglib2.0-0:amd64 (2.56.4-0ubuntu0.18.04.9) ...
Processing triggers for hicolor-icon-theme (0.17-2) ...
Processing triggers for shared-mime-info (1.9-2) ...
Errors were encountered while processing:
 python3-keyman-config
 keyman
```

**Comment by ermshiperete** [comment]
@MakaraSok The errors you get in the `dpkg` step are kind of expected - you get them because `dpkg` doesn't automatically install all dependencies python3-keyman-config requires, and so that package can't be completely installed yet. However, the next command in the instructions (`sudo apt install -f`) will fix this.

@keymanapp-test-bot retest GROUP_BIONIC TEST_KB_INSTALL1

**Comment by bharanidharanj** [comment]
### GROUP_BIONIC: Ubuntu 18.04 Bionic with Gnome Shell and X11

* **TEST_KB_INSTALL1 (FAILED):** Re-tested this in Bionic and I am getting more than 1 output in the Terminal. Attached the log file for reference. Seems to be an issue.

[km-config.log](https://github.com/keymanapp/keyman/files/8251121/km-config.log)

* **TEST_KB_INSTALL2 (FAILED):** Re-tested this in Bionic and I am getting more than 1 output in the Terminal. Attached the log file for reference. 

[km-config2.log](https://github.com/keymanapp/keyman/files/8251125/km-config2.log)

**Comment by bharanidharanj** [comment]
### GROUP_WASTA: Wasta 20.04 with Cinnamon

* **TEST_KB_INSTALL1 (OPEN):** Installed Keyman15.0.212-beta (from PR) and it is working as expected.
* **TEST_KB_INSTALL2 (OPEN):** Installed Keyman15.0.212-beta (from PR) and it is working as expected.

**Comment by bharanidharanj** [comment]
### GROUP_FOCAL: Ubuntu 20.04 Focal with Gnome Shell and X11

* **TEST_KB_INSTALL1 (PASSED):** Installed 15.0.212-beta build in Ubuntu Focal Linux and it is working as expected.
* **TEST_KB_INSTALL2 (PASSED):** Installed 15.0.212-beta build in Ubuntu Focal Linux and it is working as expected.

**Comment by bharanidharanj** [comment]
### GROUP_BIONIC: Ubuntu 18.04 Bionic with Gnome Shell and X11

* **TEST_KB_INSTALL1 (PASSED):** Installed Keyman 15.0.212-beta and re-tested this in Bionic and it is working as expected.

* **TEST_KB_INSTALL2 (PASSED):** Installed Keyman 15.0.212-beta and re-tested this in Bionic and it is working as expected.

**Comment by bharanidharanj** [comment]
### GROUP_IMPISH_X11: Ubuntu 21.10 Impish with Gnome Shell and X11

* **TEST_KB_INSTALL1 (PASSED):** Installed Keyman 15.0.212-beta build (from PR) in Impish(Ubuntu) Linux and it is working as expected. 
* **TEST_KB_INSTALL2 (PASSED):** Installed Keyman 15.0.212-beta build (from PR) in Impish(Ubuntu) Linux and it is working as expected.

### GROUP_IMPISH_WAYLAND: Ubuntu 21.10 Impish with Gnome Shell and Wayland

* **TEST_KB_INSTALL1 (PASSED):** Installed Keyman 15.0.212-beta build (from PR) in Impish (Ubuntu on Xorg)Linux and it is working as expected.
* **TEST_KB_INSTALL2 (PASSED):** Installed Keyman 15.0.212-beta build (from PR) in Impish (Ubuntu on Xorg) Linux and it is working as expected.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.213-beta](https://keyman.com/downloads/releases/beta/15.0.213)

**Pull request merged**